consensus/types/src/light_client_header.rs

```diff
@@ -1,26 +1,143 @@
-use crate::test_utils::TestRandom;
 use crate::BeaconBlockHeader;
+use crate::ForkName;
+use crate::ForkVersionDeserialize;
+use crate::{light_client_update::*, BeaconBlockBody};
+use crate::{
+    test_utils::TestRandom, EthSpec, ExecutionPayloadHeaderCapella, ExecutionPayloadHeaderDeneb,
+    FixedVector, Hash256, SignedBeaconBlock,
+};
 use serde::{Deserialize, Serialize};
+use serde_json;
 use ssz_derive::{Decode, Encode};
+use std::marker::PhantomData;
+use superstruct::superstruct;
 use test_random_derive::TestRandom;
 
-#[derive(
-    Debug,
-    Clone,
-    PartialEq,
-    Serialize,
-    Deserialize,
-    Encode,
-    Decode,
-    TestRandom,
-    arbitrary::Arbitrary,
+#[superstruct(
+    variants(Altair, Capella, Deneb),
+    variant_attributes(
+        derive(
+            Default,
+            Debug,
+            Clone,
+            PartialEq,
+            Serialize,
+            Deserialize,
+            Encode,
+            Decode,
+            TestRandom,
+            arbitrary::Arbitrary,
+        ),
+        serde(bound = "E: EthSpec", deny_unknown_fields),
+        arbitrary(bound = "E: EthSpec"),
+    )
 )]
-pub struct LightClientHeader {
+#[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, arbitrary::Arbitrary, PartialEq)]
+#[serde(untagged)]
+#[serde(bound = "E: EthSpec", deny_unknown_fields)]
+#[ssz(enum_behaviour = "union")]
```

Comment:
Why do we use `union` here?  It adds a one byte variant selector, and i think we want to use `transparent` instead?
https://github.com/sigp/ethereum_ssz/blob/649ed02ef19c6b2c364d0f4b514cc1c14f6b44d0/ssz_derive/src/lib.rs#L9

Reply:
I used `union` because the compiler was complaining about not being able to dervie `Decode` for `transparent` enum behaviours

Reply:
I see, I think I've done that before too lol but the generated implementation won't be able to decode the light client data (SSZ) received from peers, as they aren't encoded with the one-byte variant selector.  In this case we'll need to decode manually without the `Decode` derive.

I think we'll need something similar to how we handle `BeaconBlock` received over gossip - using the fork digest from the topic to determine the fork name, an example topic looks like this: `/eth2/e1925f3b/beacon_block/ssz_snappy`, where `e1925f3b` is the fork digest and it tells us the fork of this topic, so we should be able to decode using this information.

For example:
https://github.com/sigp/lighthouse/blob/4273004bd905341f6a3aa3ecbe203ce4b66e6497/beacon_node/lighthouse_network/src/types/pubsub.rs#L173-L204

and the LightClient decoding section below needs to be updated similarly:
https://github.com/sigp/lighthouse/blob/4273004bd905341f6a3aa3ecbe203ce4b66e6497/beacon_node/lighthouse_network/src/types/pubsub.rs#L267-L282

because `LightClientOptimisticUpdate` and `LightClientFinalityUpdate` aren't superstructs, you'll have to implement a function to manually decode the inner `LightClientHeader` using the fork name, a similar example here:

https://github.com/sigp/lighthouse/blob/a403138ed02f981143f41d53c24e2cae0af07baa/common/eth2/src/types.rs#L1757-L1781

Reply:
ahh makes sense, thanks!